File: FoodRegistry/src/Restaurant.cpp
#include "Restaurant.h"
#include <iostream>
#include <iomanip>
#include <cmath>
using namespace std;

Restaurant::Restaurant() {
    // Initialize menu items
    menu.push_back(MenuItem('N', "Nasi Lemak", 9.00));
    menu.push_back(MenuItem('C', "Chicken Rice", 8.00));
    menu.push_back(MenuItem('M', "Masala Dosa", 6.00));
    menu.push_back(MenuItem('H', "Hamburger", 5.00));
    menu.push_back(MenuItem('F', "Fish and Chips", 12.00));

    currentOrder = new Order(menu.size());
    dailySales = new Order(menu.size());
    customerCount = 0;
}

Restaurant::~Restaurant() {
    delete currentOrder;
    delete dailySales;
}

void Restaurant::displayWelcome() {
    cout << "= * === * === * === * === * === * ==" << endl;
    cout << "H & S Restaurant Meal Billing System" << endl;
    cout << "= * === * === * === * === * === * ==" << endl << endl;
}

void Restaurant::displayHelp() {
    cout << "=== HELP " << setw(27) << setfill('=') << "" << endl;
    cout << "List of codes for internal commands:" << endl << endl;
    cout << CMD_CLEAR << " - Clears items in the current bill." << endl;
    cout << CMD_MENU << " - Displays the menu of available items." << endl;
    cout << CMD_HELP << " - Displays this panel." << endl;
    cout << CMD_END << " - Ends the current session and generates a sales report." << endl;
    cout << setw(36) << setfill('=') << "" << endl;
    cout << "How to use:" << endl << endl;
    cout << "Input item code -> quantity" << endl;
    cout << "E.g. 'H' followed by '5' adds 5 Hamburgers" << endl;
    cout << "('-5' removes 5 Hamburgers)" << endl << endl;
    cout << "After confirming the bill," << endl;
    cout << "input the payment received in RM to checkout and close the bill." << endl;
    cout << setw(36) << setfill('=') << "" << endl;
}

void Restaurant::displayMenu() {
    cout << endl << "=== MENU " << setw(27) << setfill('=') << "" << endl << endl;
    for (const auto& item : menu) {
        item.display();
    }
    cout << setw(36) << setfill('=') << "" << endl;
}

bool Restaurant::isValidNumber(const string& str, bool allowDecimal) {
    try {
        float val = stof(str);
        if (allowDecimal) {
            return abs(val) < 524288;
        } else {
            // Check for whole number
            size_t pos = str.find('.');
            if (pos != string::npos) {
                string decimal = str.substr(pos + 1);
                for (char c : decimal) {
                    if (c != '0') return false;
                }
            }
            return abs(val) < 8192;
        }
    } catch (...) {
        return false;
    }
}

int Restaurant::findMenuItemByCode(char code) {
    for (size_t i = 0; i < menu.size(); i++) {
        if (toupper(menu[i].getCode()) == toupper(code)) {
            return i;
        }
    }
    return -1;
}

void Restaurant::processItemCode(char code) {
    int index = findMenuItemByCode(code);
    if (index == -1) {
        cout << "Unrecognised code! Simply input '?' to seek help." << endl;
        return;
    }

    string input;
    cout << "Enter the quantity of " << menu[index].getName() << " to add: ";
    cin >> input;

    if (isValidNumber(input, false)) {
        int quantity = stoi(input);
        if (abs(quantity) < 8192) {
            currentOrder->addItem(index, quantity);
            currentOrder->calculateTotals(menu);
        } else {
            cout << "Sorry, the system does not process massive quantities." << endl;
            cout << "Only input quantities less than 8192." << endl;
        }
    } else {
        cout << "Action cancelled: Item quantity should be integers." << endl;
    }
}

void Restaurant::processPayment(float amount) {
    if (amount < currentOrder->getTotal()) {
        cout << "Insufficient payment received!" << endl;
        cout << "Payment should be at least RM " << fixed 
             << setprecision(2) << currentOrder->getTotal() << endl;
        return;
    }

    customerCount++;
    cout << "Order: #" << customerCount << endl;
    currentOrder->displaySummary(menu, "Items in bill:");
    receiptPrinter.printCheckout(*currentOrder, amount, customerCount);
    receiptPrinter.saveToFile(*currentOrder, menu, amount, customerCount);

    // Add to daily sales
    for (size_t i = 0; i < menu.size(); i++) {
        dailySales->addItem(i, currentOrder->getQuantity(i));
    }
    dailySales->calculateTotals(menu);

    currentOrder->clear();
}

void Restaurant::generateDailyReport() {
    cout << endl << setw(40) << setfill('=') << "" << endl;
    cout << " DAILY MEAL SALES REPORT" << endl;
    cout << setw(40) << setfill('=') << "" << endl;
    cout << "Orders billed: " << customerCount << endl;
    dailySales->displaySummary(menu, "Items sold today:");

    cout << endl << setw(36) << setfill('=') << "" << endl;
    cout << " Subtotal: RM " << fixed << setprecision(2) << dailySales->getSubtotal() << endl;
    cout << " SST charge 10%: RM " << dailySales->getSST() << endl;
    cout << " Grand total: RM " << dailySales->getTotal() << endl;
    cout << setw(36) << setfill('=') << "" << endl;
    cout << "Thank you and have a nice day ahead!" << endl;
}

void Restaurant::run() {
    displayWelcome();
    displayHelp();
    displayMenu();

    bool running = true;
    while (running) {
        cout << endl << setw(36) << setfill('=') << "" << endl << endl;
        cout << "Order: #" << customerCount + 1 << endl;
        currentOrder->displaySummary(menu, "Items in bill:");

        cout << endl << "Enter a code to edit bill/view informations," << endl;
        cout << "or an amount in RM to checkout: ";

        string input;
        cin >> input;
        cout << endl;

        if (isValidNumber(input, true)) {
            float amount = stof(input);
            processPayment(amount);
        } else {
            char cmd = toupper(input[0]);

            switch (cmd) {
                case '<': // Clear
                    currentOrder->clear();
                    break;
                case ',': // Menu
                    displayMenu();
                    break;
                case '?': // Help
                    displayHelp();
                    break;
                case '>': // End session
                    running = false;
                    break;
                default:
                    processItemCode(cmd);
                    break;
            }
        }
    }

    generateDailyReport();
    cout << "Enter anything to exit." << endl;
    string dummy;
    cin >> dummy;
}

File: java_src/com/foodregistry/Restaurant.java
package com.foodregistry;

import java.util.ArrayList;
import java.util.List;

public class Restaurant implements IRestaurant {
    private List<MenuItem> menu;
    private Order currentOrder;
    private Order dailySales;
    private List<Order> orderHistory;
    private int customerCount;
    private Receipt receiptPrinter;

    public Restaurant() {
        menu = new ArrayList<>();
        menu.add(new MenuItem('N', "Nasi Lemak", 9.00f));
        menu.add(new MenuItem('C', "Chicken Rice", 8.00f));
        menu.add(new MenuItem('M', "Masala Dosa", 6.00f));
        menu.add(new MenuItem('H', "Hamburger", 5.00f));
        menu.add(new MenuItem('F', "Fish and Chips", 12.00f));

        currentOrder = new Order(menu.size());
        dailySales = new Order(menu.size());
        orderHistory = new ArrayList<>();
        customerCount = 0;
        receiptPrinter = new Receipt();
    }

    @Override
    public void processItemCode(char code, int quantity) {
        int index = findMenuItemByCode(code);
        if (index != -1) {
            currentOrder.addItem(index, quantity);
            currentOrder.calculateTotals(menu);
        }
    }

    private int findMenuItemByCode(char code) {
        for (int i = 0; i < menu.size(); i++) {
            if (Character.toUpperCase(menu.get(i).getCode()) == Character.toUpperCase(code)) {
                return i;
            }
        }
        return -1;
    }

    @Override
    public String processPayment(float amount) {
        if (amount < currentOrder.getTotal()) {
            return "Insufficient payment! Need RM " + String.format("%.2f", currentOrder.getTotal());
        }

        customerCount++;
        receiptPrinter.saveToFile(currentOrder, menu, amount, customerCount);
        
        // Add to history
        orderHistory.add(new Order(currentOrder));
        
        // Add to daily sales
        for (int i = 0; i < menu.size(); i++) {
            dailySales.addItem(i, currentOrder.getQuantity(i));
        }
        dailySales.calculateTotals(menu);

        String receipt = generateReceiptString(amount);
        currentOrder.clear();
        return receipt;
    }

    private String generateReceiptString(float amountPaid) {
        StringBuilder sb = new StringBuilder();
        float change = amountPaid - currentOrder.getTotal();
        sb.append("<h3>Receipt Order: #").append(customerCount).append("</h3>");
        sb.append("<div class='receipt'>");
        sb.append("Items in bill:<br>");
        for (int i = 0; i < menu.size(); i++) {
            if (currentOrder.getQuantity(i) > 0) {
                sb.append(String.format("(%c) [RM %.2f] %s x %d = RM %.2f<br>",
                    menu.get(i).getCode(), menu.get(i).getPrice(), menu.get(i).getName(),
                    currentOrder.getQuantity(i), menu.get(i).getPrice() * currentOrder.getQuantity(i)));
            }
        }
        sb.append("<hr>");
        sb.append(String.format("Subtotal: RM %.2f<br>", currentOrder.getSubtotal()));
        sb.append(String.format("SST charge 10%%: RM %.2f<br>", currentOrder.getSST()));
        sb.append(String.format("Grand total: RM %.2f<br>", currentOrder.getTotal()));
        sb.append("<hr>");
        sb.append(String.format("Amount received: RM %.2f<br>", amountPaid));
        if (change > 0) {
            sb.append(String.format("Change given: RM %.2f<br>", change));
        }
        sb.append("<hr>Thank you and have a nice day ahead!</div>");
        return sb.toString();
    }

    @Override
    public void clearOrder() {
        currentOrder.clear();
    }

    @Override
    public String generateDailyReport() {
        StringBuilder sb = new StringBuilder();
        sb.append("<h3>DAILY MEAL SALES REPORT</h3>");
        sb.append("<div class='report'>");
        
        // Transaction History
        sb.append("<h4>Transaction History</h4>");
        if (orderHistory.isEmpty()) {
            sb.append("<p>No orders processed yet.</p>");
        } else {
            for (int i = 0; i < orderHistory.size(); i++) {
                Order ord = orderHistory.get(i);
                sb.append("<div class='mini-bill' style='margin-bottom: 10px; padding-bottom: 5px; border-bottom: 1px dashed #ccc;'>");
                sb.append("<strong>Order #").append(i + 1).append("</strong><br>");
                for (int j = 0; j < menu.size(); j++) {
                    if (ord.getQuantity(j) > 0) {
                        sb.append(String.format("- %s x %d<br>", menu.get(j).getName(), ord.getQuantity(j)));
                    }
                }
                sb.append(String.format("Total: RM %.2f<br>", ord.getTotal()));
                sb.append("</div>");
            }
        }
        
        sb.append("<h4>Daily Summary</h4>");
        sb.append("Orders billed: ").append(customerCount).append("<br>");
        sb.append("Items sold today:<br>");
        
        for (int i = 0; i < menu.size(); i++) {
            if (dailySales.getQuantity(i) > 0) {
                sb.append(String.format("(%c) [RM %.2f] %s x %d = RM %.2f<br>",
                    menu.get(i).getCode(), menu.get(i).getPrice(), menu.get(i).getName(),
                    dailySales.getQuantity(i), menu.get(i).getPrice() * dailySales.getQuantity(i)));
            }
        }
        sb.append("<hr>");
        sb.append(String.format("Subtotal: RM %.2f<br>", dailySales.getSubtotal()));
        sb.append(String.format("SST charge 10%%: RM %.2f<br>", dailySales.getSST()));
        sb.append(String.format("Grand total: RM %.2f<br>", dailySales.getTotal()));
        sb.append("</div>");
        return sb.toString();
    }

    @Override
    public String getMenuDisplay() {
        StringBuilder sb = new StringBuilder();
        sb.append("<div class='menu-grid'>");
        
        for (MenuItem item : menu) {
            sb.append("<div class='menu-card'>");
            
            // Header: Code and Price
            sb.append("<div class='card-header'>");
            sb.append("<div class='item-code'>").append(item.getCode()).append("</div>");
            sb.append("<div class='item-price'>RM ").append(String.format("%.2f", item.getPrice())).append("</div>");
            sb.append("</div>");
            
            // Item Name
            sb.append("<div class='item-name'>").append(item.getName()).append("</div>");
            
            // Controls Form
            sb.append("<form action='/action' method='post' class='item-controls'>");
            sb.append("<input type='hidden' name='action' value='add'>");
            sb.append("<input type='hidden' name='code' value='").append(item.getCode()).append("'>");
            
            sb.append("<div class='qty-wrapper'>");
            sb.append("<button type='button' class='btn-qty minus' onclick='updateQty(this, -1)'>-</button>");
            sb.append("<input type='number' name='quantity' value='0' min='1' class='qty-input' readonly>");
            sb.append("<button type='button' class='btn-qty plus' onclick='updateQty(this, 1)'>+</button>");
            sb.append("</div>");
            
            sb.append("<button type='submit' class='btn-add'>Add to Cart</button>");
            sb.append("</form>");
            
            sb.append("</div>"); // end card
        }
        
        sb.append("</div>");
        return sb.toString();
    }

    @Override
    public String getCurrentOrderDisplay() {
        if (currentOrder.isEmpty()) return "<p>Nothing.</p>";
        
        StringBuilder sb = new StringBuilder();
        sb.append("<div class='order-list'>");
        for (int i = 0; i < menu.size(); i++) {
            if (currentOrder.getQuantity(i) > 0) {
                sb.append(String.format("<div>(%c) %s x %d = RM %.2f</div>",
                    menu.get(i).getCode(), menu.get(i).getName(), 
                    currentOrder.getQuantity(i), 
                    menu.get(i).getPrice() * currentOrder.getQuantity(i)));
            }
        }
        sb.append("<hr>");
        sb.append(String.format("<div><strong>Amount to be paid: RM %.2f</strong></div>", currentOrder.getTotal()));
        sb.append("</div>");
        return sb.toString();
    }
}
